File: backend/workflow/saved.go
package workflow

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/enough/enough/backend/enoughhome"
)

type SavedWorkflow struct {
	Name    string
	Path    string
	Project bool
	Meta    Meta
}

var savedNamePattern = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{0,63}$`)

func ScanSaved(workDir string) []SavedWorkflow {
	byName := map[string]SavedWorkflow{}
	homeRoot := filepath.Join(enoughhome.HomeDir(), "workflows", "saved")
	scanSavedRoot(homeRoot, false, byName)
	scanSavedRoot(filepath.Join(workDir, ".enough", "workflows", "saved"), true, byName)
	out := make([]SavedWorkflow, 0, len(byName))
	for _, item := range byName {
		out = append(out, item)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	return out
}

func scanSavedRoot(root string, project bool, out map[string]SavedWorkflow) {
	entries, err := os.ReadDir(root)
	if err != nil {
		return
	}
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		name := entry.Name()
		path := filepath.Join(root, name, "workflow.js")
		if _, err := os.Stat(path); err != nil {
			continue
		}
		item := SavedWorkflow{Name: name, Path: path, Project: project}
		if data, err := os.ReadFile(filepath.Join(root, name, "meta.json")); err == nil {
			_ = json.Unmarshal(data, &item.Meta)
		}
		out[name] = item
	}
}

func SaveWorkflow(scriptPath, name, workDir string, project bool) (SavedWorkflow, error) {
	name = strings.ToLower(strings.TrimSpace(name))
	if !savedNamePattern.MatchString(name) {
		return SavedWorkflow{}, errors.New("workflow name must use lowercase letters, digits, and hyphens")
	}
	data, err := os.ReadFile(scriptPath)
	if err != nil {
		return SavedWorkflow{}, err
	}
	root := filepath.Join(enoughhome.HomeDir(), "workflows", "saved")
	if project {
		root = filepath.Join(workDir, ".enough", "workflows", "saved")
	}
	dir := filepath.Join(root, name)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return SavedWorkflow{}, err
	}
	dst := filepath.Join(dir, "workflow.js")
	if err := os.WriteFile(dst, data, 0o644); err != nil {
		return SavedWorkflow{}, err
	}
	meta, _ := Inspect(scriptPath)
	meta.Name = name
	metaData, _ := json.MarshalIndent(meta, "", "  ")
	if err := os.WriteFile(filepath.Join(dir, "meta.json"), metaData, 0o644); err != nil {
		return SavedWorkflow{}, err
	}
	return SavedWorkflow{Name: name, Path: dst, Project: project, Meta: meta}, nil
}

File: backend/workflow/saved.ts
// PORT: backend/workflow/saved.go

import { Effect } from "effect";
import fs from "node:fs";
import path from "node:path";
import { home_dir } from "../hollowhome/home";
import { type Meta } from "./types";
import { Inspect } from "./runtime";

export interface SavedWorkflow {
  Name: string;
  Path: string;
  Project: boolean;
  Meta: Meta;
}

const savedNamePattern = /^[a-z0-9][a-z0-9-]{0,63}$/;

export function ScanSaved(workDir: string): SavedWorkflow[] {
  const byName: Record<string, SavedWorkflow> = {};
  const homeRoot = path.join(home_dir(), "workflows", "saved");
  scanSavedRoot(homeRoot, false, byName);
  scanSavedRoot(path.join(workDir, ".hollow", "workflows", "saved"), true, byName);

  const out: SavedWorkflow[] = Object.values(byName);
  out.sort((a, b) => a.Name.localeCompare(b.Name));
  return out;
}

function scanSavedRoot(root: string, project: boolean, out: Record<string, SavedWorkflow>): void {
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(root, { withFileTypes: true });
  } catch {
    return;
  }
  for (const entry of entries) {
    if (!entry.isDirectory()) {
      continue;
    }
    const name = entry.name;
    const wfPath = path.join(root, name, "workflow.js");
    try {
      if (!fs.statSync(wfPath).isFile()) {
        continue;
      }
    } catch {
      continue;
    }
    const item: SavedWorkflow = {
      Name: name,
      Path: wfPath,
      Project: project,
      Meta: { name: "", description: "" },
    };
    try {
      const metaData = fs.readFileSync(path.join(root, name, "meta.json"), "utf8");
      item.Meta = JSON.parse(metaData);
    } catch {}
    out[name] = item;
  }
}

export function SaveWorkflow(
  scriptPath: string,
  name: string,
  workDir: string,
  project: boolean
): Effect.Effect<SavedWorkflow, Error> {
  return Effect.tryPromise({
    try: async () => {
      const normalizedName = name.trim().toLowerCase();
      if (!savedNamePattern.test(normalizedName)) {
        throw new Error("workflow name must use lowercase letters, digits, and hyphens");
      }

      let data: Buffer;
      try {
        data = fs.readFileSync(scriptPath);
      } catch (err: any) {
        throw new Error(`read script ${scriptPath}: ${err.message}`);
      }

      const root = project
        ? path.join(workDir, ".hollow", "workflows", "saved")
        : path.join(home_dir(), "workflows", "saved");

      const dir = path.join(root, normalizedName);
      try {
        fs.mkdirSync(dir, { recursive: true, mode: 0o755 });
      } catch (err: any) {
        throw new Error(`mkdir ${dir}: ${err.message}`);
      }

      const dst = path.join(dir, "workflow.js");
      try {
        fs.writeFileSync(dst, data, { mode: 0o644 });
      } catch (err: any) {
        throw new Error(`write ${dst}: ${err.message}`);
      }

      let meta: Meta = { name: normalizedName, description: "" };
      try {
        const metaRes = await Effect.runPromise(Inspect(scriptPath));
        meta = { ...metaRes, name: normalizedName };
      } catch {}

      try {
        const metaData = JSON.stringify(meta, null, "  ");
        fs.writeFileSync(path.join(dir, "meta.json"), metaData, { mode: 0o644 });
      } catch (err: any) {
        throw new Error(`write meta.json: ${err.message}`);
      }

      return { Name: normalizedName, Path: dst, Project: project, Meta: meta };
    },
    catch: (cause) => (cause instanceof Error ? cause : new Error(String(cause))),
  });
}

/*
PORT STATUS
source path: backend/workflow/saved.go
source lines: 88
draft lines: 113
confidence: high
status: phase_b_compile
*/
